applications/tari_dan_app_utilities/src/template_manager/implementation/cmap_semaphore.rs

```diff
@@ -0,0 +1,67 @@
+//   Copyright 2023 The Tari Project
+//   SPDX-License-Identifier: BSD-3-Clause
+
+use std::{
+    fmt::{Debug, Formatter},
+    hash::Hash,
+    sync::{Arc, Mutex, MutexGuard},
+};
+
+#[derive(Clone)]
+pub struct CMapSemaphore<K: Hash + Eq> {
+    map: Arc<dashmap::DashMap<K, Arc<Mutex<()>>>>,
+    global: Arc<std_semaphore::Semaphore>,
+}
+
+impl<K: Hash + Eq + Clone> CMapSemaphore<K> {
+    pub fn new(max_global_access: isize) -> Self {
+        Self {
+            map: Arc::new(dashmap::DashMap::new()),
+            global: Arc::new(std_semaphore::Semaphore::new(max_global_access)),
+        }
+    }
+
+    pub fn acquire(&self, key: K) -> CMapSemaphoreGuard<'_, K> {
+        let global_access = self.global.access();
+        let map_mutex = self
+            .map
+            .entry(key.clone())
+            .or_insert_with(|| Arc::new(Mutex::new(())))
+            .clone();
+        CMapSemaphoreGuard {
+            _global_access: global_access,
+            map: self.map.clone(),
+            map_mutex,
+            key,
+        }
+    }
+}
+
+impl<K: Hash + Eq> Debug for CMapSemaphore<K> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CMapSemaphore")
+            .field("map", &self.map.len())
+            .field("global", &"...")
+            .finish()
+    }
+}
+
+pub struct CMapSemaphoreGuard<'a, K: Hash + Eq> {
+    _global_access: std_semaphore::SemaphoreGuard<'a>,
```

Comment:
Perhaps add a comment here to prevent people from deleting this thinking it is unused